src/osx/core/colour.cpp

```diff
@@ -13,127 +13,244 @@
+    virtual CGFloat Green() const wxOVERRIDE { return m_green; }
+    virtual CGFloat Blue() const wxOVERRIDE { return m_blue; }
+    virtual CGFloat Alpha() const wxOVERRIDE { return m_alpha; }
+
+    CGColorRef GetCGColor() const wxOVERRIDE { return m_cgColour; }
+
+    virtual wxColorRefData* Clone() const wxOVERRIDE { return new wxCGColorRefData(*this); }
+
+private:
+    void Init(CGFloat components[4]);
+
+    wxCFRef<CGColorRef> m_cgColour;
+
+    CGFloat m_red;
+    CGFloat m_blue;
+    CGFloat m_green;
+    CGFloat m_alpha;
+
+    wxDECLARE_NO_ASSIGN_CLASS(wxCGColorRefData);
+};
+
+wxCGColorRefData::wxCGColorRefData(CGFloat r, CGFloat g, CGFloat b, CGFloat a)
+{
+    CGFloat components[4];
+    components[0] = r;
+    components[1] = b;
+    components[2] = g;
+    components[3] = a;
+
+    Init(components);
+}
+
+wxCGColorRefData::wxCGColorRefData(CGFloat components[4])
+{
+    Init(components);
+}
+
+wxCGColorRefData::wxCGColorRefData(const wxCGColorRefData& other)
+{
+    m_red = other.m_red;
+    m_blue = other.m_blue;
+    m_green = other.m_green;
+    m_alpha = other.m_alpha;
+
+    m_cgColour = other.m_cgColour;
+}
+
+void wxCGColorRefData::Init(CGFloat components[4])
+{
+    m_red = components[0];
+    m_blue = components[1];
+    m_green = components[2];
+    m_alpha = components[3];
+
+    CGColorRef cfcol;
+    cfcol = CGColorCreate(wxMacGetGenericRGBColorSpace(), components);
+
+    wxASSERT_MSG(cfcol != NULL, "Invalid CoreGraphics Color");
+    m_cgColour.reset(cfcol);
+}
+
+wxCGColorRefData::wxCGColorRefData(CGColorRef col)
+{
+    wxASSERT_MSG(col != NULL, "Invalid CoreGraphics Color");
+    m_cgColour.reset(col);
+
+    wxCFRef<CGColorRef> rgbacol;
+    size_t noComp = CGColorGetNumberOfComponents(col);
+    const CGFloat* components = CGColorGetComponents(col);
+
+    // set default alpha
+    m_alpha = 1.0;
+    bool isRGB = true;
+
+    CGColorSpaceModel model = CGColorSpaceGetModel(CGColorGetColorSpace(col));
+    if (model == kCGColorSpaceModelMonochrome)
+    {
+        wxASSERT_MSG(1 <= noComp && noComp <= 2, "Monochrome Color unexpected components");
+        m_red = components[0];
+        m_green = components[0];
+        m_blue = components[0];
+        if (noComp > 1)
+            m_alpha = components[1];
+        isRGB = false;
+    }
+    else if (model != kCGColorSpaceModelRGB)
+    {
+#if MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_11
+        if (wxPlatformInfo::Get().CheckOSVersion(10, 11))
+        {
+            rgbacol = CGColorCreateCopyByMatchingToColorSpace(wxMacGetGenericRGBColorSpace(), kCGRenderingIntentDefault, col, NULL);
+            noComp = CGColorGetNumberOfComponents(rgbacol);
+            components = CGColorGetComponents(rgbacol);
+        }
+        else
+#endif
+        {
+            isRGB = false;
+        }
+    }
+
+    if (isRGB)
+    {
+        wxASSERT_MSG(3 <= noComp && noComp <= 4, "Monochrome Color unexpected components");
```

Comment:
I don't understand this assert message, why is it monochrome?

Reply:
error on my part